monitoring/exporter/exporter.go

```diff
@@ -0,0 +1,232 @@
+	// returned, and in that case the error is reported to callers via OnError and the row data
+	// will not be uploaded to stackdriver. When GetProjectID is not set, all row data will be
+	// considered not applicable to this exporter.
+	GetProjectID func(*RowData) (projectID string, err error)
+	// OnError is used to report any error happened while exporting view data fails. Whenever
+	// this function is called, it's guaranteed that at least one row data is also passed to
+	// OnError. Row data passed to OnError must not be modified. When OnError is not set, all
+	// errors happened on exporting are ignored.
+	OnError func(error, ...*RowData)
+	// MakeResource creates monitored resource from RowData. It is guaranteed that only RowData
+	// that passes GetProjectID will be given to this function. Though not recommended, error
+	// can be returned, and in that case the error is reported to callers via OnError and the
+	// row data will not be uploaded to stackdriver. When MakeResource is not set, global
+	// resource is used for all RowData objects.
+	MakeResource func(rd *RowData) (*monitoredrespb.MonitoredResource, error)
+
+	// options concerning labels.
+
+	// DefaultLabels store default value of some labels. Labels in DefaultLabels need not be
+	// specified in tags of view data. Default labels and tags of view may have overlapping
+	// label keys. In this case, values in tag are used. Default labels are used for labels
+	// those are constant throughout export operation, like version number of the calling
+	// program.
+	DefaultLabels map[string]string
+	// UnexportedLabels contains key of labels that will not be exported stackdriver. Typical
+	// uses of unexported labels will be either that marks project ID, or that's used only for
+	// constructing resource.
+	UnexportedLabels []string
+}
+
+// default values for options
+func defaultGetProjectID(rd *RowData) (string, error) {
+	return "", RowDataNotApplicableError
+}
+
+func defaultOnError(err error, rds ...*RowData) {}
+
+func defaultMakeResource(rd *RowData) (*monitoredrespb.MonitoredResource, error) {
+	return &monitoredrespb.MonitoredResource{Type: "global"}, nil
+}
+
+// NewStatsExporter creates a StatsExporter object. Once a call to NewStatsExporter is made, any
+// fields in opts must not be modified at all. ctx will also be used throughout entire exporter
+// operation when making RPC call.
+func NewStatsExporter(ctx context.Context, opts *Options) (*StatsExporter, error) {
+	client, err := newMetricClient(ctx, opts.ClientOptions...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a metric client: %v", err)
+	}
+
+	e := &StatsExporter{
+		ctx:         ctx,
+		client:      client,
+		opts:        opts,
+		projDataMap: make(map[string]*projectData),
+	}
+
+	// We don't want to modify user-supplied options, so save default options directly in
+	// exporter.
+	if opts.GetProjectID != nil {
```

Comment:
You can replace all these if-else blocks with an initialization followed by an if-block.

Example:
	e.onError = defaultOnError
	if opts.OnError != nil {
		e.onError = opts.OnError
	}

Reply:
ok